### PR TITLE
Truffle environment

### PR DESCRIPTION
**Pull request by fainashalts**
This PR seeks to move environment-related files to a `truffle-environment` package so that they can be more easily required from outside of the `truffle-core` package. In doing so, the webpack config for the path to `chain.js` has also been changed.

Environment-related code currently exists in the code for several Truffle commands, so these were updated to required the new path.

+ update some code to use `const` and `let` rather than `var` where appropriate

### CODE REVIEW

File: packages/truffle-environment/package.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "truffle-environment",
+  "version": "1.0.0",
+  "description": "This package contains the code needed to determine environment-specific data",
+  "main": "index.js",
+  "author": "Faina Shalts",
```
Comment (CruzMolina) [inline packages/truffle-environment/package.json]: I think the dependencies for the pkg need to be added here (i.e. `ganache-core`, `node-ipc`, `web3`, etc)

Comment (CruzMolina) [reply packages/truffle-environment/package.json]: Oh, also believe we want the initial pkg version to be `"0.1.0"`

File: packages/truffle-environment/index.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+const environment = require("./environment");
+const develop = require("./develop");
+
+module.exports = { environment, develop };
```
Comment (gnidan) [inline packages/truffle-environment/index.js]: These should keep the capitalized names

File: packages/truffle-core/lib/commands/migrate.js [hunk 3]
```diff
@@ -134,7 +134,7 @@ const command = {
     const Resolver = require("truffle-resolver");
     const Migrate = require("truffle-migrate");
     const Config = require("truffle-config");
-    const Environment = require("../environment");
+    const Environment = require("truffle-environment").environment;
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/migrate.js]:
```suggestion
    const { Environment } = require("truffle-environment");
```

File: packages/truffle-core/lib/commands/migrate.js [hunk 4]
```diff
@@ -167,7 +167,7 @@ const command = {
     const Resolver = require("truffle-resolver");
     const Migrate = require("truffle-migrate");
     const Contracts = require("truffle-workflow-compile");
-    const Environment = require("../environment");
+    const Environment = require("truffle-environment").environment;
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/migrate.js]:
```suggestion
    const { Environment } = require("truffle-environment");
```

File: packages/truffle-core/lib/commands/test.js [hunk 5]
```diff
@@ -47,12 +47,12 @@ const command = {
     const temp = require("temp").track();
     const Config = require("truffle-config");
     const Artifactor = require("truffle-artifactor");
-    const Develop = require("../develop");
+    const Develop = require("truffle-environment").develop;
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/test.js]:
```suggestion
    const { Environment, Develop } = require("truffle-environment");
```

File: packages/truffle-core/lib/commands/test.js [hunk 6]
```diff
@@ -47,12 +47,12 @@ const command = {
...
     const Test = require("../test");
     const fs = require("fs");
     const { promisify } = require("util");
     const promisifiedCopy = promisify(require("../copy"));
-    const Environment = require("../environment");
+    const Environment = require("truffle-environment").environment;
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/test.js]:
```suggestion
```

File: packages/truffle-core/lib/commands/console.js [hunk 7]
```diff
@@ -19,21 +19,21 @@ var command = {
     ]
   },
   run: function(options, done) {
-    var Config = require("truffle-config");
-    var Console = require("../console");
-    var Environment = require("../environment");
+    const Config = require("truffle-config");
+    const Console = require("../console");
+    const Environment = require("truffle-environment").environment;
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/console.js]:
```suggestion
    const { Environment } = require("truffle-environment");
```

File: packages/truffle-core/lib/commands/debug.js [hunk 8]
```diff
@@ -21,7 +21,7 @@ const command = {
     const debug = debugModule("lib:commands:debug");
 
     const Config = require("truffle-config");
-    const Environment = require("../environment");
+    const Environment = require("truffle-environment").environment;
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/debug.js]:
```suggestion
    const { Environment } = require("truffle-environment");
```

File: packages/truffle-core/lib/commands/develop.js [hunk 9]
```diff
@@ -16,7 +16,7 @@ const command = {
   },
   runConsole: (config, ganacheOptions, done) => {
     const Console = require("../console");
-    const Environment = require("../environment");
+    const Environment = require("truffle-environment").environment;
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/develop.js]:
```suggestion
    const { Environment } = require("truffle-environment");
```

File: packages/truffle-core/lib/commands/develop.js [hunk 10]
```diff
@@ -43,7 +43,7 @@ const command = {
   },
   run: (options, done) => {
     const Config = require("truffle-config");
-    const Develop = require("../develop");
+    const Develop = require("truffle-environment").develop;
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/develop.js]:
```suggestion
    const { Develop } = require("truffle-environment");
```

File: packages/truffle-core/lib/commands/exec.js [hunk 11]
```diff
@@ -36,17 +36,17 @@ var command = {
...
-    var ConfigurationError = require("../errors/configurationerror");
-    var Require = require("truffle-require");
-    var Environment = require("../environment");
-    var path = require("path");
-    var OS = require("os");
+    const Config = require("truffle-config");
+    const Contracts = require("truffle-workflow-compile");
+    const ConfigurationError = require("../errors/configurationerror");
+    const Require = require("truffle-require");
+    const Environment = require("truffle-environment").environment;
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/exec.js]:
```suggestion
    const { Environment } = require("truffle-environment");
```

File: packages/truffle-core/lib/commands/migrate.js (hunk 3 above)
Comment (fainashalts) [reply packages/truffle-core/lib/commands/migrate.js]: love this! will make the change! actually meant to ask you about a prettier way to do this. :)

File: packages/truffle-environment/package.json [hunk 12]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "truffle-environment",
+  "version": "0.1.0",
+  "description": "This package contains the code needed to determine environment-specific data",
+  "main": "index.js",
+  "author": "Faina Shalts",
+  "license": "MIT",
+  "private": false,
+  "dependencies": {
+    "debug": "^4.1.0",
```
Comment (CruzMolina) [inline packages/truffle-environment/package.json]: Just a tiny thing but I think `debug` can be a `devDep`

File: packages/truffle-environment/package.json [hunk 13]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "truffle-environment",
+  "version": "0.1.0",
+  "description": "This package contains the code needed to determine environment-specific data",
+  "main": "index.js",
+  "author": "Faina Shalts",
+  "license": "MIT",
```
Comment (CruzMolina) [inline packages/truffle-environment/package.json]: Oh, can we get a LICENSE file added to the pkg?

File: packages/truffle-environment/package.json [hunk 14]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "truffle-environment",
+  "version": "0.1.0",
+  "description": "This package contains the code needed to determine environment-specific data",
+  "main": "index.js",
```
Comment (CruzMolina) [inline packages/truffle-environment/package.json]: Might be nice to add a [basic simple README](https://github.com/trufflesuite/truffle/tree/develop/packages/truffle-blockchain-utils#readme) to the pkg and [`"repository"`](https://github.com/trufflesuite/truffle/blob/develop/packages/truffle-workflow-compile/package.json#L19), [`"bugs"/"url"`](https://github.com/trufflesuite/truffle/blob/develop/packages/truffle-workflow-compile/package.json#L22-L24), and [`"homepage"`](https://github.com/trufflesuite/truffle/blob/develop/packages/truffle-workflow-compile/package.json#L25) key/value entries to the `package.json`

File: packages/truffle-environment/package.json [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+  "dependencies": {
+    "debug": "^4.1.0",
+    "ganache-core": "2.5.5",
+    "node-ipc": "^9.1.1",
+    "truffle-artifactor": "^4.0.19",
+    "truffle-error": "^0.0.5",
+    "truffle-expect": "^0.0.9",
+    "truffle-interface-adapter": "^0.1.6",
+    "truffle-resolver": "^5.0.14",
+    "truffle-workflow-compile": "^2.0.19",
```
Comment (CruzMolina) [inline packages/truffle-environment/package.json]: Not sure if we need `truffle-workflow-compile` added to the pkg.

File: packages/truffle-environment/package.json [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    "universal-analytics": "^0.4.17",
```
Comment (CruzMolina) [inline packages/truffle-environment/package.json]: Same with `universal-analytics`.

File: packages/truffle-environment/package.json (hunk 15 above)
Comment (fainashalts) [reply packages/truffle-environment/package.json]: You're totally right. Removed.

File: packages/truffle-environment/package.json (hunk 16 above)
Comment (fainashalts) [reply packages/truffle-environment/package.json]: Removed.

File: packages/truffle-environment/package.json (hunk 12 above)
Comment (fainashalts) [reply packages/truffle-environment/package.json]: moved!

File: packages/truffle-environment/package.json (hunk 13 above)
Comment (fainashalts) [reply packages/truffle-environment/package.json]: Yep!

File: packages/truffle-environment/package.json (hunk 14 above)
Comment (fainashalts) [reply packages/truffle-environment/package.json]: Done!

File: packages/truffle-environment/package.json [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+  "name": "truffle-environment",
+  "version": "0.1.0",
+  "description": "This package contains the code needed to determine environment-specific data",
+  "main": "index.js",
+  "author": "Faina Shalts",
+  "license": "MIT",
+  "repository": "https://github.com/trufflesuite/truffle/tree/master/packages/truffle-environment",
+  "homepage": "https://github.com/trufflesuite/truffle-environment#readme",
+  "bugs": {
+    "url": "https://github.com/trufflesuite/truffle-environment/issues"
```
Comment (CruzMolina) [inline packages/truffle-environment/package.json]:
```suggestion
    "url": "https://github.com/trufflesuite/truffle/issues"
```

File: packages/truffle-environment/package.json [hunk 18]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "truffle-environment",
+  "version": "0.1.0",
+  "description": "This package contains the code needed to determine environment-specific data",
+  "main": "index.js",
+  "author": "Faina Shalts",
+  "license": "MIT",
+  "repository": "https://github.com/trufflesuite/truffle/tree/master/packages/truffle-environment",
+  "homepage": "https://github.com/trufflesuite/truffle-environment#readme",
```
Comment (CruzMolina) [inline packages/truffle-environment/package.json]:
```suggestion
  "homepage": "https://github.com/trufflesuite/truffle/tree/master/packages/truffle-environment#readme",
```

**Pull request merged**